### PR TITLE
ENH: Add more const qualifiers to C API arguments

### PR DESCRIPTION
**Pull request by WillAyd**
ref https://github.com/numpy/numpy/issues/24855#issuecomment-1939313091

### CODE REVIEW

File: numpy/_core/include/numpy/ndarraytypes.h [hunk 1]
```diff
@@ -1514,13 +1514,13 @@ PyArray_STRIDE(const PyArrayObject *arr, int istride)
 }
 
 static inline NPY_RETURNS_BORROWED_REF PyObject *
-PyArray_BASE(PyArrayObject *arr)
+PyArray_BASE(const PyArrayObject *arr)
```
Comment (WillAyd) [inline numpy/_core/include/numpy/ndarraytypes.h:1517]: Wasn't sure about this one but assumed we don't want users to modify the base through this. 

On a related note I left `PyArray_BYTES` and `PyArray_DATA` as non-const out of these

Comment (seberg) [reply numpy/_core/include/numpy/ndarraytypes.h:1517]: Isn't const fine in all of these cases, since the object is const for the function call?

Comment (WillAyd) [reply numpy/_core/include/numpy/ndarraytypes.h:1517]: Good point - I'll add those as well

**Comment by ngoldbaum** [comment]
I think this will need a mention in the numpy 2.0 transition guide, since this will trigger new compiler errors for C++ code (not C though, IIRC) using these function.

**Comment by ngoldbaum** [comment]
Probably needs a release note too.

**Comment by seberg** [comment]
I checked to be sure, because C++ can be picky:  It doesn't seem a problem here at least with `static inline` function.

A brief release note snippet would be nice either way.

**Comment by mattip** [comment]
On the one hand these are harmless because the input is cast from `const ...` to a different non-const type. But doesn't that mean they also do not really "protect" the user from modifying the input since the returned value is not `const`?

**Comment by seberg** [comment]
The only point is allowing the user to use `const PyArrayObject *` without compiler warnings if they prefer.  (I _think_ that is even the only change.)

**Review by seberg — APPROVED** [review]

**Comment by ngoldbaum** [comment]
I just double-checked and this shouldn’t lead to new compiler warnings, so no need to include a release note either. Feel free to add one in a followup if you feel like it. Please open a bug if we’re missing a corner case and this change breaks your code somehow.

**Pull request merged**